plotter/general_test.go

```diff
@@ -0,0 +1,128 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package plotter
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/vg"
+)
+
+var generateTestData = flag.Bool("regen", false, "Uses the current state to regenerate the test data.")
+
+// checkPlot checks a generated plot against a previously created reference.
+// If generateTestData = true, it regenerates the reference first.
+func checkPlot(dir, name, ext string, p *plot.Plot, width, height vg.Length,
+	errf errFunc, lf logFunc) {
+
+	filename := filepath.Join(dir, fmt.Sprintf("%s.%s", name, ext))
+
+	c, err := p.WriterTo(width, height, ext)
+	errf(err)
+
+	var buf bytes.Buffer
+	_, err = c.WriteTo(&buf)
+	errf(err)
+
+	// Recreate Golden images.
+	if *generateTestData {
+		errf(p.Save(width, height, filename))
+	}
+
+	f, err := os.Open(filename)
+	errf(err)
+
+	want, err := ioutil.ReadAll(f)
+	errf(err)
+	f.Close()
+	if !bytes.Equal(buf.Bytes(), want) {
+		errf(fmt.Errorf("image mismatch for %s\n", filename))
+		return
+	}
+	lf("Image can be seen at https://github.com/gonum/plot/tree/master/plotter/%s/%s.%s.\n"+
+		"Normally, you would use plot.Save().\n", dir, name, ext)
+}
+
+// errFunc is an interface for functions that deal with errors.
+type errFunc func(err error)
+
+// handleEx prints err if err!=nil
+var handleEx = func(err error) {
+	if err != nil {
+		fmt.Println(err)
+	}
+}
+
+// handleTest returns a function that fails the test if err!=nil
+var handleTest = func(t *testing.T) errFunc {
+	return func(err error) {
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+type logFunc func(format string, args ...interface{})
+
+var testLog = func(t *testing.T) logFunc {
+	return func(s string, arg ...interface{}) {
+		t.Logf(s, arg...)
+	}
+}
+
+var exampleLog = func(s string, arg ...interface{}) {
```

Comment:
Another option would be to create wrappers for checkPlot:

``` go
func checkPlotExample(dir, name, ext string, p *plot.Plot, width, height vg.Length) {
	checkPlot(dir, name, ext, p, width, height,
		func(err error) {
			if err != nil {
				fmt.Println(err)
			}
		},
		func(s string, arg ...interface{}) {
			fmt.Printf(s, arg...)
		},
	)
}

func checkPlotTest(dir, name, ext string, p *plot.Plot, width, height vg.Length, t *testing.T) {
	checkPlot(dir, name, ext, p, width, height,
		func(t *testing.T) errFunc {
			return func(err error) {
				if err != nil {
					t.Fatal(err)
				}
			}
		}(t),
		func(t *testing.T) logFunc {
			return func(s string, arg ...interface{}) {
				t.Logf(s, arg...)
			}
		}(t),
	)
}

```

I'm not sure the extra complexity is worth it to avoid the copypasting though.